### ISSUE TITLE
 ask an Idea ... without repeating a code I did in admin registration?

### ISSUE THREAD

**Issue by WK-GiHu** [Justinjay282/qwerty#46]
Can I also ask an Idea sir? As you can see, I've created an admin system in python. It includes the user and admin registration. In some cases, both the user and the admin have the same input. Is there a way to it efficiently? without repeating a code I did in admin registration?
Just like this:
*Admin_registration*:
*GUI HERE*  store data to the database `user_db`
*User_registration*
Since the info is the same as admin
I just copied the GUI. but the database name is `admin_db`
Is there a way to write another code efficiently without copying what I did in the admin registration?

_Originally posted by @Justinjay282 in https://github.com/Justinjay282/qwerty/issues/45#issuecomment-611863162_

**Comment by WK-GiHu** [comment]
@Justinjay282 

>Is there a way to write another code efficiently without copying what I did in the admin registration?

If you are sure, the table name are the only difference, pass the table name, e.g. `'residents_db'`, as a argument.
```
class Registration (tk.Toplevel):
    def __init__(self, parent, tabel_name):
        self.table_name = table_name
```

If you need logical differences also, define the above class as BASE class.
```
class ResidentsRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='residents_db')

**Comment by Justinjay282** [comment]
sir @WK-GiHu, 
my Registration GUI is in `class Registration`? So it'll be like this?
GUI in `class Registration` (*GUI of both admin and user*) there's a `button` submit. I have a `def submit_register admin(...` if I clicked admin, I'll have `def registered_admin(...` then if the user will register then I'll proceed to code block below? 
```
class ResidentsRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='residents_db')
```

**Comment by WK-GiHu** [comment]
@Justinjay282 

>if the user will register then I'll proceed to code block below?

No, the code block below is the definition, you have to instantiate a object accordingly if you want **admin** or **residents**.
```
registering = AdminRegistration(parent=self)

# OR

registering = ResidentsRegistration(parent=self)
```

If you decide to use inheritance, your **admin** definition looks like:
```
class AdminRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='admin_db')
```

**Comment by Justinjay282** [comment]
sir @WK-GiHu, 
so my GUI is in
```
class Registration (tk.Toplevel):
    def __init__(self, parent, tabel_name):
        self.table_name = table_name
```
I'll inherit from `class Registration...` 
```
class AdminRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='admin_db')

class ResidentRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='admin_db')
```

I'm still confused about this code block
```
registering = AdminRegistration(parent=self)

# OR

registering = ResidentsRegistration(parent=self)
```
I'm sorry since I really don't have an idea and I'm not experienced doing inheritance I really have no clue...

**Comment by Justinjay282** [comment]
> sir @WK-GiHu,
> so my GUI is in
> 
> ```
> class Registration (tk.Toplevel):
>     def __init__(self, parent, tabel_name):
>         self.table_name = table_name
> ```
> 
> I'll inherit from `class Registration...`
> 
> ```
> class AdminRegistration(Registration):
>     def __init__(self, parent):
>         super ().__init__(parent, tabel_name='admin_db')
> 
> class ResidentRegistration(Registration):
>     def __init__(self, parent):
>         super ().__init__(parent, tabel_name='admin_db')
> ```
> 
> I'm still confused about this code block
> 
> ```
> registering = AdminRegistration(parent=self)
> 
> # OR
> 
> registering = ResidentsRegistration(parent=self)
> ```
> 
> I'm sorry since I really don't have an idea and I'm not experienced doing inheritance I really have no clue...

**Comment by WK-GiHu** [comment]
@Justinjay282 

>what is this line `super ().__init__(parent, tabel_name='admin_db')`.

This init the BASE class, it defines which table_name this object have to use. Here, you define **all** differences between `admin` and `residents`. 

---

>When I created an admin system, I have `Frame.__init__(self, controller)`

I see
https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/admin_system_controller#L12
please point me to the proper file.

---

>what are the properties in pythozn?

It's a abbreviation of 
- class members
- class variables 
- class instance variables
- class attributes

In Python, there is a decorator called
```
@property
```
which is used to access a class method like a class variable or to define a read only class member name.

---

> I'm still not experiencing what are these.

Read again [you have to ...](https://github.com/Justinjay282/qwerty/issues/46#issuecomment-612040812)

**Comment by Justinjay282** [comment]
sir @WK-GiHu 
> 
> > what is this line `super ().__init__(parent, tabel_name='admin_db')`.
> 
> This init the BASE class, it defines which table_name this object have to use. Here, you define **all** differences between `admin` and `residents`.
> 
> > When I created an admin system, I have `Frame.__init__(self, controller)`
> 
> I see
> https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/admin_system_controller#L12

> 
> 
> please point me to the proper file.

https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/admin_system_controller#L55
Here sir, I have `Frame.__init__(self, parent)`

> > what are the properties in python?

> 
> It's a abbreviation of
> 
> * class members
> * class variables
> * class instance variables
> * class attributes
> 
> In Python, there is a decorator called
> 
> ```
> @property
> ```
> 
> which is used to access a class method like a class variable or to define a read only class member name.
> 
> > I'm still not experiencing what are these.
> 
> Read again [you have to ...](https://github.com/Justinjay282/qwerty/issues/46#issuecomment-612040812)

I now understand:
```
registering = AdminRegistration(parent=self)

# OR

registering = ResidentsRegistration(parent=self)
```
with this I can call the `AdminRegistration(parent=self)`  as registering.

`registering.submit_registration()`

But I'm still confused.

Can you check if this is correct? 
```
class Registration (tk.Toplevel):
    def __init__(self, parent, table_name):
        self.table_name = table_name
    
    def submit_registration(self):
          #store to database
          ...

class ResidentsRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='residents_db')
        Registration.submit_registration()
        
class AdminRegistration(Registration):
    def __init__(self, parent):
        super ().__init__(parent, tabel_name='admin_db')
        # Additional GUI need to be filled up by the admin
        Registration.submit_registration()
```

**Comment by WK-GiHu** [comment]
@Justinjay282 

>with this I can call the `AdminRegistration(parent=self)` as registering.

Not call!
This statement instantiate a object `AdminRegistration` named `registering` and executes the class `__init__` method.

---

>check if this is correct?
>`Registration.submit_registration()`

You have defined a class Methode, therefore you can't call it using @classmethod syntax. You have to use: `self.submit_registration(...`.
```
def submit_registration(self):
```
---

I don't get why you want to call `submit_registration` in `__init"__`?
Here you have to do **nothing**, you have to extend `class Registration.__init__` with all you are doing in
https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/admin_system_controller#L13
You have to restrict to one `TABLE` which name is passed with argument `table_name=`.
https://github.com/Justinjay282/qwerty/blob/4aafa475ca3473d796c4983feb396288d5c1d999/admin_system_controller#L18-L21

**Comment by Justinjay282** [comment]
sir @WK-GiHu, It's still confusing to me.

Since I can't understand it, I've put it on code and this is my progress so far. I've created a new file, using Tkinter with a controller to navigate easily named `admin_system_inheritance.py`

It's my progress so far, I'm having a problem on how I can pass values to be filled up to store in the database...

**Comment by Justinjay282** [comment]
Hi, sir @WK-GiHu, I'm converting my executable file. I already did it. But the database isn't there.
Can I ask how can I create an executable file with a pymysql database? so I don't need to install the xampp phpmyadmin?
